config/configoptional/optional.go

```diff
@@ -0,0 +1,98 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package configoptional // import "go.opentelemetry.io/collector/config/configoptional"
+
+import (
+	"go.opentelemetry.io/collector/confmap"
+)
+
+// Optional is a type that can be used to represent a value that may or may not be present.
+// It supports two flavors: Some(value), and Default(factory).
+type Optional[T any] struct {
+	value    T
+	hasValue bool
+
+	// defaultFn returns a default value for the type T.
+	// It MUST be nil if `hasValue` is true.
+	defaultFn *DefaultFunc[T]
+}
+
+// DefaultFunc is a function type that returns a default value of type T.
+type DefaultFunc[T any] func() T
+
+var _ confmap.Unmarshaler = (*Optional[any])(nil)
+
+// Some creates an Optional with a value and no factory.
+func Some[T any](value T) Optional[T] {
+	return Optional[T]{value: value, hasValue: true}
+}
+
+// Factory is an opaque type that can be used to create a default value for the type T.
+// Factories should be package variables.
+type Factory[T any] struct {
+	defaultFn *DefaultFunc[T]
+}
+
+// NewFactory creates a new Factory with the given default function.
+// Factories should be package variables, to ensure the default function
+// is the same for all instances of the type T.
+func NewFactory[T any](defaultFn DefaultFunc[T]) Factory[T] {
+	return Factory[T]{defaultFn: &defaultFn}
+}
+
+// Default creates an Optional which has no value
+// and a factory to create a default value.
+//
+// On successful unmarshal, the factory is erased.
+func Default[T any](factory Factory[T]) Optional[T] {
```

Comment:
why is factory needed? What's the difference if it just takes `DefaultFunc[T]` directly?